swiftui/PasswordlessSwiftUI/PasswordlessSwiftUI/ContentView.swift

```diff
@@ -67,11 +68,15 @@ struct ContentView: View {
   }
 
   private func passwordlessSignIn(email: String, link: String,
-                                  completion: @escaping (_ user: User?, _ error: Error?) -> ()) {
+                                  completion: @escaping (Result<User?, Error>) -> Void) {
     Auth.auth().signIn(withEmail: email, link: link) { result, error in
-      print(error ?? "Authentication was successful.")
-      let newUser = error == nil ? Auth.auth().currentUser : nil
-      completion(newUser, error)
+      if let error = error {
+        print("ⓧ Authentication error: \(error.localizedDescription).")
+        completion(.failure(error))
+      } else {
+        print("✔ Authentication was successful.")
+        completion(.success(Auth.auth().currentUser))
```

Comment:
This could be `result?.user` instead of getting the singleton again.

This doesn't need changing right now, just wanted to point out this is an example where our API could be improved - right now you're forced to return an optional User object (or you could force unwrap it) when there:should be only 2 use cases: successfully signed in, or an error.

If we wanted to handle this case fully, we could check if `result` is also `nil` and wrap our own custom error, returning it in the completion handler. Don't think it's fully necessary but thought it was a good learning opportunity.

Reply:
yes- i understand!